### DISCUSSION TITLE
How to use fx results as command arguments

### DISCUSSION THREAD

**Discussion by mviereck** [ImageMagick/ImageMagick#5953] [Help] [answered]
### ImageMagick version

7.1.0-57 Q16-HDRI

### Operating system, version and so on

debian bullseye

### Description

Is there a way to use `fx` for command arguments?

For example, I want to do something like `-level [fx:minima*100]%,[fx:maxima*100]%`.

For now I use code like:
```
min="$(magick rose: -format "%[fx:minima*100]" info:)"
max="$(magick rose: -format "%[fx:maxima*100]" info:)"
magick rose: -level $min%,$max% show:
```
A single line like this would be nice:
```
magick rose: -level [fx:minima*100]%,[fx:maxima*100]% show:
```

**Comment by snibgo** [comment, marked as answer]
With v7 `magick`, in most places that expect numbers we can use `%[fx:...]` expressions. Note the leading `%`. For example:
```
magick rose: -level %[fx:minima*1.1],%[fx:maxima*0.9] out.png
```

**Reply by mviereck** [reply]
Thank you!
Odd that I did not see this anywhere in the documentation or in an example.

I've tried a slightly different syntax and had issues.
This one gives the same result as `-auto-level`:
```
magick -size 400x400 radial-gradient: +sigmoidal-contrast 50 -level %[fx:minima*quantumrange],%[fx:maxima*quantumrange] show:
```
However, this one gives a message (no exiting error) and a wrong result:
```
$ magick -size 400x400 radial-gradient: +sigmoidal-contrast 50 -level %[fx:minima*100]%,%[fx:maxima*100]% show:
magick: unknown image property "%," @ warning/property.c/InterpretImageProperties/3881.
```
It seems that `magick` has issues to parse the second % in the first argument that is literally meant as 'percent'.
A % attached to the second argument only is accepted and gives the expected result:
```
magick -size 400x400 radial-gradient: +sigmoidal-contrast 50 -level %[fx:minima*quantumrange],%[fx:maxima*100]% show:
```

**Comment by snibgo** [comment]
See the `-level` documentation at https://imagemagick.org/script/command-line-options.php#level

Roughly speaking, the documented syntax is:
```
-level black_point,white_point%
```
... where the `%` is optional. If the `%` is present, then both numbers will be taken as percentages. Including `%` between black_point and the comma is not documented syntax.

The IM interpreter is often permissive, and will sometimes allow a `%` after black_point. But not when black_point is an `%[fx:...]` expression.

**Reply by mviereck** [reply]
I did some test runs today and now I get the expected results without errors. Thank you for your patience! Great that it is possible to use fx directly.
